### PR TITLE
TST: Fix unit tests that used to call unittest.TestCase.fail

### PR DESCRIPTION
**Pull request by rth**
Following pytest migration https://github.com/numpy/numpy/issues/10856 there are unit tests that used to call `unittest.TestCase.fail` and now will produce a misleading error message on failure because the `fail` method no longer exists (test classes just inherit from `object`).

This fixes this issue, by raising an exception on failure explicitly.

### CODE REVIEW

**Comment by charris** [comment]
`ValueError` looks wrong for most of these, for test failures `AssertionError` is expected. It also looks like most of these old tests should be rewritten to use newer tools, like `assert_raises`, `assert_raises_regex`, and such. Some of those can also be used as context managers.

**Comment by rth** [comment]
Thanks for the review @charris!

You are right `assert_raises`, etc context managers are much better. Updated the PR to address your comments. Also extended this to a few other tests that used similar mechanism.

Here I tried to use `assert_raises_regex` when the error message helped tests readability, and `assert_raises` when the exception type was specific enough (e.g. `FloatingPointError`). There are still a few places in more complicated constructs where `AssertionError` is manually raised.

File: numpy/core/tests/test_regression.py [hunk 1]
```diff
@@ -1314,9 +1314,10 @@ def test_zeros(self):
             np.empty(sz)
         except ValueError as e:
             if not str(e) == good:
-                self.fail("Got msg '%s', expected '%s'" % (e, good))
+                raise AssertionError("Got msg '%s', expected '%s'" % (e, good))
```
Comment (eric-wieser) [inline numpy/core/tests/test_regression.py]: `assert_raises_regex` would work well here

File: numpy/core/tests/test_regression.py [hunk 2]
```diff
@@ -1328,9 +1329,10 @@ def test_huge_arange(self):
             assert_(np.size == sz)
         except ValueError as e:
             if not str(e) == good:
-                self.fail("Got msg '%s', expected '%s'" % (e, good))
+                raise AssertionError("Got msg '%s', expected '%s'" % (e, good))
```
Comment (eric-wieser) [inline numpy/core/tests/test_regression.py]: And here

**Review by eric-wieser — COMMENTED** [review]
Looks pretty good, thanks - two cases you missed where `assert_raises_regex` can be used

**Comment by rth** [comment]
Thanks for the review! After the reading the code, I have not realized that that construction was a way to implement `assert_raises_regex`. Updated the PR.

**Comment by eric-wieser** [comment]
Thanks @rth!

**Pull request merged**